### DISCUSSION TITLE
gRPC Migration

### DISCUSSION THREAD

**Discussion by Emantor** [labgrid-project/labgrid#1417] [Announcements]
Hi,

@jluebbe, @Bastian-Krause, and me spent the last days deep in the remote code to migrate from the WAMP based crossbar infrastructure towards gRPC. Unfortunately we did not completely finish the migration, but have an implementation which works in principle. I will post a very rough draft PR next week.

**Comment by jluebbe** [comment]
Our current work-in-progress branch is here: https://github.com/labgrid-project/labgrid/blob/grpc/README.grpc

Note that, we'll likely rebase it again in the future.

**Reply by jluebbe** [reply]
We've tested the gRPC branch with internal testsuites and also on the LXA TAC. So far it works well. Our plan is to switch our lab to it tomorrow and see how it goes.

**Reply by jluebbe** [reply]
We've seen some minor issues that need to be fixed before considering a merge:

- [x] `labgrid-client release --kick` doens't work
- [x] `labgrid-client allow` seems broken
- [ ] the gRPC shutdown in labgrid-client causes errors with Python 3.12.3
- [x] restarting exporters can cause an inconsistent resource state
- [x] migrate containers

**Reply by Bastian-Krause** [reply]
> Note that, we'll likely rebase it again in the future.

Due to merge conflicts with latest master, I've rebased the grpc branch on master. So expect a forced update.

**Reply by Bastian-Krause** [reply]
I've pushed another forced update with clean commits rebased on master.

**Reply by Bastian-Krause** [reply]
Another force push: The grpc branch is now ready for a merge in master: see #1469.

**Comment by flxzt** [comment]
Would it make sense to make reservations part of the gRPC client stream? As far as I can see the only way to get up to date information about them is to continuously poll through `rpc GetReservations` which is not ideal for e.g. continuously running clients like a GUI..

**Reply by jluebbe** [reply]
> Would it make sense to make reservations part of the gRPC client stream? As far as I can see the only way to get up to date information about them is to continuously poll through `rpc GetReservations` which is not ideal for e.g. continuously running clients like a GUI..

Yes, but it should be on-request, perhaps for now via a new `Subscribe.kind.all_reservations`?

**Reply by flxzt** [reply]
yes that would work I think!

**Comment by brherger** [comment]
Any updates on the official release timeline?

**Reply by Bastian-Krause** [reply]
@jluebbe just tagged [v25.0](https://github.com/labgrid-project/labgrid/releases/tag/v25.0)
